operator/controllers/runtimetask_reconcile_test.go

```diff
@@ -0,0 +1,264 @@
+			if got := newTaskGroupChildProxy(tt.args.node, tt.args.tasks...); !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("newTaskGroupChildProxy() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func Test_reconcileTasks(t *testing.T) {
+	nodes := []corev1.Node{
+		{ // a node with a matching task/pending
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node1-a",
+			},
+		},
+		{ // a node with a matching task/running
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node1-b",
+			},
+		},
+		{ // a node with a matching task/completed
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node1-c",
+			},
+		},
+		{ // a node with a matching task/failed
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node1-d",
+			},
+		},
+		{ // a node without a matching task (task to be created)
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node2",
+			},
+		},
+		{ // a node with two matching tasks (invalid)
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node3",
+			},
+		},
+	}
+	tasks := &operatorv1.RuntimeTaskList{
+		Items: []operatorv1.RuntimeTask{
+			{ // a pending task matching node 1
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "task1-a",
+				},
+				Spec: operatorv1.RuntimeTaskSpec{
+					NodeName: "node1-a",
+				},
+			},
+			{ // a running task matching node 1
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "task1-b",
+				},
+				Spec: operatorv1.RuntimeTaskSpec{
+					NodeName: "node1-b",
+				},
+				Status: operatorv1.RuntimeTaskStatus{
+					StartTime: timePtr(metav1.Now()),
+				},
+			},
+			{ // a completed task matching node 1
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "task1-c",
+				},
+				Spec: operatorv1.RuntimeTaskSpec{
+					NodeName: "node1-c",
+				},
+				Status: operatorv1.RuntimeTaskStatus{
+					StartTime:      timePtr(metav1.Now()),
+					CompletionTime: timePtr(metav1.Now()),
+				},
+			},
+			{ // a failed task matching node 1
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "task1-d",
+				},
+				Spec: operatorv1.RuntimeTaskSpec{
+					NodeName: "node1-d",
+				},
+				Status: operatorv1.RuntimeTaskStatus{
+					StartTime:    timePtr(metav1.Now()),
+					ErrorMessage: stringPtr("error"),
+				},
+			},
+			{ // a task matching node 3
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "task2",
+				},
+				Spec: operatorv1.RuntimeTaskSpec{
+					NodeName: "node3",
+				},
+			},
+			{ // another task matching node 3
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "task3",
+				},
+				Spec: operatorv1.RuntimeTaskSpec{
+					NodeName: "node3",
+				},
+			},
+			{ // a task not matching any node (invalid)
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "task4",
+				},
+				Spec: operatorv1.RuntimeTaskSpec{
+					NodeName: "node4",
+				},
+			},
+		},
+	}
+
+	got := reconcileTasks(nodes, tasks)
+
+	// All
+	want := []string{"node1-a", "node1-b", "node1-c", "node1-d", "node2", "node3", "node4"}
+	if got := taskNames(got.all); !reflect.DeepEqual(got, want) {
+		t.Errorf("newTaskGroupChildProxy().all = %v, want %v", got, want)
```

Comment:
this list of "all", "pending"...etc, might be simplified with reflection and a table, but will be more difficult to read.

Reply:
"all", "pending" etc are all part of the same struct, and all of them are set by the same call. Having a table will make it really difficult to read